sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -106,21 +106,37 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     val query = plan(ctx.queryNoWith)
 
     // Apply CTEs
-    query.optional(ctx.ctes) {
-      val ctes = ctx.ctes.namedQuery.asScala.map { nCtx =>
-        val namedQuery = visitNamedQuery(nCtx)
-        (namedQuery.alias, namedQuery)
-      }
-      // Check for duplicate names.
-      checkDuplicateKeys(ctes, ctx)
-      With(query, ctes)
+    query.optionalMap(ctx.ctes)(withCTE)
+  }
+
+  private def withCTE(ctx: CtesContext, plan: LogicalPlan): LogicalPlan = {
+    val ctes = ctx.namedQuery.asScala.map { nCtx =>
+      val namedQuery = visitNamedQuery(nCtx)
+      (namedQuery.alias, namedQuery)
     }
+    // Check for duplicate names.
+    checkDuplicateKeys(ctes, ctx)
+    With(plan, ctes)
   }
 
   override def visitQueryToDesc(ctx: QueryToDescContext): LogicalPlan = withOrigin(ctx) {
     plan(ctx.queryTerm).optionalMap(ctx.queryOrganization)(withQueryResultClauses)
   }
 
+  override def visitQueryWithFrom(ctx: QueryWithFromContext): LogicalPlan = withOrigin(ctx) {
+    val from = visitFromClause(ctx.fromClause)
+    validate(ctx.querySpecification.fromClause == null,
+      "Script transformation queries cannot have a FROM clause in their" +
```

Comment:
Am I missing anything, is this related to script transformation?

Reply:
@viirya First of all you are absolutely right. I could use your suggestion here. So what this is guarding against is of query of following form : 
```
FROM foo select c1 from foo
```
So this has nothing to do with script transformation :-). I don't know what to call this kind of query as ? Should we say "Individual select statement can not have FROM cause as its already specified in the outer block" ?

Reply:
`Individual select statement can not have FROM cause as its already specified in the outer query block` SGTM

Reply:
@cloud-fan Thank you. will change.

Reply:
SGTM too.

Reply:
Thank you @viirya